src/gt4py/next/program_processors/runners/dace_fieldview/transformations/map_fusion_helper.py

```diff
@@ -200,367 +194,475 @@ def relocate_nodes(
-        if set(params_1) != set(params_2):
-            return False
 
-        # Maps the name of a parameter to the dimension index
-        param_dim_map_1: dict[str, int] = {pname: i for i, pname in enumerate(params_1)}
-        param_dim_map_2: dict[str, int] = {pname: i for i, pname in enumerate(params_2)}
+        # The parameter names
+        first_params: List[str] = first_map.params
+        second_params: List[str] = second_map.params
+
+        if len(first_params) != len(second_params):
+            return None
+
+        # The ranges, however, we apply some post processing to them.
+        simp = lambda e: symbolic.simplify_ext(symbolic.simplify(e))  # noqa: E731
+        first_rngs: Dict[str, Tuple[Any, Any, Any]] = {
+            param: tuple(simp(r) for r in rng) for param, rng in zip(first_params, first_map.range)
+        }
+        second_rngs: Dict[str, Tuple[Any, Any, Any]] = {
+            param: tuple(simp(r) for r in rng)
+            for param, rng in zip(second_params, second_map.range)
+        }
+
+        # Parameters of the second map that have not yet been matched to a parameter
+        #  of the first map and vice versa.
+        unmapped_second_params: Set[str] = set(second_params)
+        unused_first_params: Set[str] = set(first_params)
+
+        # This is the result (`second_param -> first_param`), note that if no renaming
+        #  is needed then the parameter is not present in the mapping.
+        final_mapping: Dict[str, str] = {}
+
+        # First we identify the parameters that already have the correct name.
+        for param in set(first_params).intersection(second_params):
+            first_rng = first_rngs[param]
+            second_rng = second_rngs[param]
+
+            if first_rng == second_rng:
+                # They have the same name and the same range, this is already a match.
+                #  Because the names are already the same, we do not have to enter them
+                #  in the `final_mapping`
+                unmapped_second_params.discard(param)
+                unused_first_params.discard(param)
+
+        # Check if no remapping is needed.
+        if len(unmapped_second_params) == 0:
+            return {}
+
+        # Now we go through all the parameters that we have not mapped yet.
+        #  All of them will result in a remapping.
+        for unmapped_second_param in unmapped_second_params:
+            second_rng = second_rngs[unmapped_second_param]
+            assert unmapped_second_param not in final_mapping
+
+            # Now look in all not yet used parameters of the first map which to use.
+            for candidate_param in unused_first_params:
+                candidate_rng = first_rngs[candidate_param]
+                if candidate_rng == second_rng:
+                    final_mapping[unmapped_second_param] = candidate_param
+                    unused_first_params.discard(candidate_param)
+                    break
+            else:
```

Comment:
In Python the `for` loop looks like
```python
for iter_var in iter_space:
    loop-body
else:
  else-body
```

The `else` is entered if the `for` ended normally, i.e. it iterated over everything in `iteration_space`, but it will not be entered if you exit the loop through `break`.
You could emulate this with an `if` and boolean flag.

Reply:
This is new for me, nice!